### ISSUE TITLE
Extend Delve backend RPC API to be able to make a Delve REPL front-end console in a IDE

### ISSUE THREAD

**Issue by gertcuykens** [derekparker/delve#383]
Some commands are better in a console interface, to create a console in a IDE it would be much easier if the API has also REPL front-end support.

Example: https://github.com/Microsoft/vscode-go/issues/186

Please discuss with @lukehoban I am just the ticket man :)

**Comment by lukehoban** [comment]
Thanks @gertcuykens. 

To provide some more technical details on the issue here:

IDE integrations make calls through the RPC API, so ensure that they can programmatically invoke commands and read results in a structured way.  

However, in some cases, the IDE integration may want to support also allowing users to drop into a REPL where they can send commands to the current instance of `dlv` using the same front-end commands as the `dlv` binary.

Right now, the structure of Delve doesn't appear to make this possible.  There is no way to invoke the Delve front-end code (`terminal/command.go`) from the RPC API (`server/rpc/client.go`).  This design is nicely decoupled.  But it doesn't support this IDE scenario (unless the IDE effectively re-implements `terminal/command` and stays up to date with it).

**Comment by aarzilli** [comment]
Would it be unacceptable to run a dlv frontend with `dlv connect` and communicate with it using pipes? 
It seems that the biggest problem is that the API is not reentrant, especially `Command`, so you would have to make sure no two calls are ever made simultaneously.

**Comment by lukehoban** [comment]
@aarzilli 

Good suggestion - connecting with `dlv connect` would be an acceptable solution I think.

Unfortunately, it doesn't seem to work when connecting to a headless server that is in the middle of a debug session that has been accepting RPC requests from another client.  I tried this, and I can connect, but any commands I try to execute hang `dlv`.

```
c:\dd\go\src\github.com\lukehoban\hello>dlv connect 127.0.0.1:2345
Type 'help' for list of commands.
(dlv) sources

```

If I then continue the other client to the end of the debug session, the `dlv connect` session unhangs with:

```
Command failed: read tcp 127.0.0.1:4991->127.0.0.1:2345: wsarecv: An existing connection was forcibly closed by the remote host.
(dlv)
```

Is this expected?

**Comment by aarzilli** [comment]
It seems the server only accepts one connection, probably the correct behavior give how the API works. I've added a flag to allow multiple clients in PR #395.

**Comment by aarzilli** [comment]
Closed by c277b2715750a46153b2974f66eec84d438f669c
